rover_gui/src/QDeviceStatus/QDeviceStatus.cpp

```diff
@@ -0,0 +1,270 @@
+ *
+ * @param deviceID_
+ */
+void QDeviceStatus::updateRebootCounter(RoverCan2::Constant::eDeviceId deviceID_)
+{
+    if (!_deviceInfo.contains(deviceID_) || !_deviceReboot.contains(deviceID_))
+    {
+        return;
+    }
+
+    uint16_t& deviceReboots = _numberOfDeviceReboots[deviceID_];
+    uint16_t& oldDeviceReboots = _oldDeviceReboots[deviceID_];
+    uint16_t& deviceRebootsFromButton = _numberOfDeviceRebootsFromButton[deviceID_];
+    uint16_t& deviceMessageCount = _deviceMessageCount[deviceID_];
+
+    if (_numberOfCalls < deviceMessageCount)
+    {
+        oldDeviceReboots = 0;
+        deviceReboots = 0;
+        return;
+    }
+
+    deviceReboots = deviceRebootsFromButton + _numberOfCalls - deviceMessageCount;
+
+    if (deviceReboots != oldDeviceReboots)
+    {
+        oldDeviceReboots = deviceReboots;
+        uint16_t deviceID = static_cast<uint16_t>(deviceID_);
+        RCLCPP_INFO(rclcpp::get_logger("GUI"), "Device %d has rebooted since last call", deviceID);
+    }
+}
+
+/**
+ * @brief Updates the label text with the updated status report for the received device.
+ *
+ * @param deviceID_
+ */
+void QDeviceStatus::setStatusReport(RoverCan2::Constant::eDeviceId deviceID_)
+{
+    if (!_deviceInfo.contains(deviceID_) || !_deviceReboot.contains(deviceID_))
+    {
+        return;
+    }
+
+    QLabel* label = _deviceReboot[deviceID_];
+
+    std::string deviceName = this->getDeviceName(deviceID_);
+    uint16_t deviceID = static_cast<uint16_t>(deviceID_);
+
+    QString labelText = QString::fromStdString(deviceName) + QString("\n\nID: 0x%1").arg(deviceID, 3, 16, QChar('0'))
+                        + "\n\nReboots: " + QString::number(_numberOfDeviceReboots[deviceID_]);
+
+    label->setText(labelText);
+}
+
+/**
+ * @brief Updates the label color based on the device status. If ok: green, if warning: yellow, if error: red.
+ *
+ * @param deviceID_
+ * @param deviceStatus_
+ */
+void QDeviceStatus::updateDeviceColor(RoverCan2::Constant::eDeviceId deviceID_, const rover_msgs::msg::CanDeviceStatus& msg_)
+{
+    if (!_deviceInfo.contains(deviceID_) || !_deviceReboot.contains(deviceID_))
+    {
+        return;
+    }
+
+    QWidget* widgetInfo = _deviceInfo[deviceID_];
+
+    switch (msg_.error_state)
+    {
+        case rover_msgs::msg::CanDeviceStatus::STATUS_OK:
+            widgetInfo->setStyleSheet(STATUS_SUCCESS);
+            break;
+        case rover_msgs::msg::CanDeviceStatus::STATUS_WARNING:
+            widgetInfo->setStyleSheet(STATUS_WARNING);
+            break;
+        case rover_msgs::msg::CanDeviceStatus::STATUS_ERROR:
+            widgetInfo->setStyleSheet(STATUS_ERROR);
+            break;
+        default:
+            widgetInfo->setStyleSheet(STATUS_DEFAULT);
+            break;
+    }
+}
+
+/**
+ * @brief Sets the default style for all device labels.
+ *
+ */
+void QDeviceStatus::setDefaultStyle()
+{
+    for (auto& it : _deviceInfo)
+    {
+        it.second->setStyleSheet(STATUS_DEFAULT);
+    }
+}
```

Comment:
nitpick: for cpp23 it's prefered to use structured binding declaration:


```suggestion
    for (auto& [key, value] : _deviceInfo)
    {
        value->setStyleSheet(STATUS_DEFAULT);
    }
}
```